### ISSUE TITLE
Support Java 17

### ISSUE THREAD

**Issue by maxandersen** [quarkusio/quarkus#21481]
### Description
Java 17 is out and Quarkus JVM works with Java 17 but native image needs updating and devtools/code quarkus not fully covered. 

Using this epic to track status

### Analysis
(links to analysis docs containing architecture design work, requirements gathering, etc)

**Comment by famod** [comment]
One question here would be about CI (or nightly jobs). Do we need more Java 17 coverage? It seems so.

**Comment by maxandersen** [comment]
good question @famod - what java 17 coverage do we actually have enabled in CI now ?

**Comment by famod** [comment]
The following modules are _not_ executed with Java 17:
- integration-tests/devtools
- integration-tests/gradle
- integration-tests/maven
- tcks

Plus: No native tests, AFAICS, but I'm not really up to date with which GraalVM version we are running currently.

Ironically, the nightly EA job (using 18-ea currently) is covering all those (except native).
PS: Seems I forgot to create a houskeeping issue after #20340.

**Comment by zakkak** [comment]
@famod regarding native-image testing I am planning to create pipelines for testing Quarkus (running its integration tests) with Mandrel on Java 17. This would give us some info, but it would of course not ensure that no breaking changes make it to Quarkus moving forward.

> Plus: No native tests, AFAICS, but I'm not really up to date with which GraalVM version we are running currently.

We are currently running with 21.3-java11

**Comment by zakkak** [comment]
FYI all [native integration tests are now green with Mandrel 21.3-java17](https://github.com/graalvm/mandrel/actions/runs/1531746358).

Note, the `--add-exports` added to resolve the issues in native mode:

* https://github.com/quarkusio/quarkus/pull/21886
* https://github.com/quarkusio/quarkus/pull/21785
* https://github.com/quarkusio/quarkus/pull/21408
* https://github.com/quarkusio/quarkus/pull/21550

will probably be needed in JVM-mode to avoid unpleasant surprises once the APIs in those modules get accessed at runtime.

**Comment by famod** [comment]
Devtools ITs are now also running with Java 17 in CI.

**Comment by zakkak** [comment]
@maxandersen You might want to include https://github.com/oracle/graal/issues/4006 as well

**Comment by michael-simons** [comment]
Records do fail also with Qute. No warnings / errors on build time, but on runtime

```
Caused by: io.quarkus.qute.TemplateException: Property "value" not found on the base object "de.jcvogt.girlswhoviking.Quiz$Answer" in expression {foo.value} in template index.qute.html on line 24
```

when applying `@RegisterForReflection`, same error as in https://github.com/quarkusio/quarkus/issues/20891

**Comment by vukzeka** [comment]
I agree with @zakkak this list should include [oracle/graal#4006](https://github.com/oracle/graal/issues/4006)

**Comment by maxandersen** [comment]
All issues here are resolved and closed.  We support Java 17 for a while now - anything else found is considered a bug.
